### PR TITLE
Prefer KnownEndpointsUrl, to EndpointsUrl, to local default

### PR DESCRIPTION
**Pull request by boblangley**
Connects to https://github.com/Particular/ServiceInsight/issues/774

We have introduced a distinct API call to return the known endpoints from ServiceControl which is the, now, more correct call for ServiceInsight to make.

The change is intended to be backwards compatible, falling back to the "endpoints" const string when no other service is reported by ServiceControl for SI to use.

### CODE REVIEW

File: src/ServiceInsight/ServiceControl/DefaultServiceControl.cs [hunk 1]
```diff
@@ -134,7 +141,16 @@ public IEnumerable<StoredMessage> GetConversationById(string conversationId)
...
         public IEnumerable<Endpoint> GetEndpoints()
         {
-            var request = new RestRequestWithCache(EndpointsEndpoint, RestRequestWithCache.CacheStyle.IfNotModified);
+            var rootUrls = GetRootUrls();
+
+            var endpointsUrl = rootUrls.KnownEndpointsUrl ?? rootUrls.EndpointsUrl ?? DefaultEndpointsEndpoint;
+
+            if (endpointsUrl.Contains(connection.Url))
+            {
+                endpointsUrl = endpointsUrl.Replace(connection.Url, string.Empty);
```
Comment (danielmarbach) [inline src/ServiceInsight/ServiceControl/DefaultServiceControl.cs]: could we just use replace without the contains?

File: src/ServiceInsight/ServiceControl/DefaultServiceControl.cs [hunk 2]
```diff
@@ -81,6 +81,13 @@ public string GetVersion()
             return header == null ? null : header.Value.ToString();
         }
 
+        ServiceControlRootUrls GetRootUrls()
```
Comment (danielmarbach) [inline src/ServiceInsight/ServiceControl/DefaultServiceControl.cs]: @Particular/serviceinsight-maintainers this API could be used to replace all the hardcoded URIs in the future. We figured that is something that can be done at a later stage as a refactoring

File: src/ServiceInsight/ServiceControl/DefaultServiceControl.cs [hunk 3]
```diff
@@ -134,7 +141,16 @@ public IEnumerable<StoredMessage> GetConversationById(string conversationId)
 
         public IEnumerable<Endpoint> GetEndpoints()
         {
-            var request = new RestRequestWithCache(EndpointsEndpoint, RestRequestWithCache.CacheStyle.IfNotModified);
+            var rootUrls = GetRootUrls();
+
+            var endpointsUrl = rootUrls.KnownEndpointsUrl ?? rootUrls.EndpointsUrl ?? DefaultEndpointsEndpoint;
```
Comment (adamralph) [inline src/ServiceInsight/ServiceControl/DefaultServiceControl.cs]: As discussed, we will put the Elvis operator into the `rootUrls` calls.

File: src/ServiceInsight/ServiceControl/DefaultServiceControl.cs (hunk 1 above)
Comment (adamralph) [reply src/ServiceInsight/ServiceControl/DefaultServiceControl.cs]: As discussed: yes.

**Comment by adamralph** [comment]
Marking as WIP because this needs some more smoke tests.

**Comment by adamralph** [comment]
Pushed a fix up with the changes we discussed. Still a WIP while we smoke test and before we squash.

**Comment by SeanFeldman** [comment]
@boblangley and I have fixed an issue when SC was returning KnownEndpointView rather than EndpointView (older API). Smoked tested with this PR and a fallback (again `master`). Both render endpoints as expected. @danielmarbach @adamralph please review.

**Comment by danielmarbach** [comment]
@SeanFeldman @boblangley heads up I rebased this on top of the async changes introduced in #781  You need to reset your branches locally

**Comment by WilliamBZA** [comment]
Is this still WIP?

File: src/ServiceInsight/Models/Endpoint.cs [hunk 4]
```diff
@@ -2,23 +2,31 @@
 {
     public class Endpoint
     {
-        string host;
-
-        public string Name { get; set; }
+        public string Name
+        {
+            get => EndpointDetails.Name;
```
Comment (HEskandari) [inline src/ServiceInsight/Models/Endpoint.cs]: This uses expression assignment instead of old get/set style. Not that I don't like it but for consistency, I think we need to keep using the older/established coding style. Same goes for below.

File: src/ServiceInsight/Models/Endpoint.cs [hunk 5]
```diff
@@ -2,23 +2,31 @@
...
+        public string HostId
+        {
+            get => EndpointDetails.HostId;
+            set => EndpointDetails.HostId = value;
+        }
 
         public string HostDisplayName { get; set; }
 
-        public string Address => string.Format("{0}{1}", Name, AtMachine());
+        public string Address => $"{Name}{AtMachine()}";
```
Comment (HEskandari) [inline src/ServiceInsight/Models/Endpoint.cs]: This does not seem to be related to the task at hand. Suggesting to keep as separate refactoring commit/PR.

File: src/ServiceInsight/Models/Endpoint.cs [hunk 6]
```diff
@@ -2,23 +2,31 @@
...
+
+        public EndpointDetails EndpointDetails { get; set; } = new EndpointDetails();
 
-        string AtMachine() => string.IsNullOrEmpty(Host) ? string.Empty : string.Format("@{0}", Host);
+        string AtMachine() => string.IsNullOrEmpty(Host) ? string.Empty : $"@{Host}";
```
Comment (HEskandari) [inline src/ServiceInsight/Models/Endpoint.cs]: This does not seem to be related to the task at hand. Suggesting to keep as separate refactoring commit/PR.

**Comment by danielmarbach** [comment]
Not from my side. I tested this locally but I think @adamralph marked it as wip for more testing

File: src/ServiceInsight/Models/Endpoint.cs (hunk 6 above)
Comment (danielmarbach) [reply src/ServiceInsight/Models/Endpoint.cs]: @HEskandari why? The class was touched in this PR. While not ideal I don't see why we can't leave it as is

File: src/ServiceInsight/Models/Endpoint.cs (hunk 4 above)
Comment (adamralph) [reply src/ServiceInsight/Models/Endpoint.cs]: Personaly I'd rather see us move everything gradually toward expression bodied members. We should probably be consistent within a single file, though, and I think we are here.

Comment (adamralph) [reply src/ServiceInsight/Models/Endpoint.cs]: Discussed this with @danielmarbach and we think this is better, and it is consistent with other existing members in this file.

File: src/ServiceInsight/Models/Endpoint.cs (hunk 5 above)
Comment (adamralph) [reply src/ServiceInsight/Models/Endpoint.cs]: Discussed with @danielmarbach and we agree that this is minor enough that we can leave as-is this time.

File: src/ServiceInsight/Models/Endpoint.cs (hunk 6 above)
Comment (adamralph) [reply src/ServiceInsight/Models/Endpoint.cs]: Discussed with @danielmarbach and we agree that this is minor enough that we can leave as-is this time.
